### PR TITLE
Use address format from server in address card

### LINKED ISSUE #42106: Use `get_address_formats` to format AddressCard instead of hardcoded method.
WooCommerce core provides a way to format addresses based on a country, this formatting lives in the [`WC_Countries::get_address_formats`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-countries.php#L553) and is filterable via `woocommerce_localisation_address_formats`. We should ideally use that instead of hardcoding the address ourselves.

Just a note that uses 2 things that we need to handle:
- It uses new line breaks instead of commas, we can handle that on the component level by replacing `\n` with `,`.
- It relies on tokens to search-replace values, we need to handle that and normalize for it.
    - The amount of tokens differs a bit, for example, some tokens have a `_upper` postfix which means the word need to be capitalized, sometimes the first and last names are compained into a single one `{name}`
- Alternatively, use the default method.

### LINKED ISSUE #43500: Name order not respected by the address card block on the checkout page
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

In Hungary and Japan, we use a different name order. This is already handled here:

https://github.com/woocommerce/woocommerce/blob/c487c84778b02946ce811f89368f369d282f1d3e/plugins/woocommerce/includes/class-wc-countries.php#L574

And here:

https://github.com/woocommerce/woocommerce/blob/c487c84778b02946ce811f89368f369d282f1d3e/plugins/woocommerce/includes/class-wc-countries.php#L1154-L1161

(same with Japan).

### Expected behavior

Based on this, the name on the address card should be reversed too if the customer has a hungarian or japanese address.

### Actual behavior

However at the moment this name order is only respected on the checkout form, non on the address card.

On the form the order is correct:
Screenshot 2024-01-11 at 10 54 48

On the address card, the name is reversed:
Screenshot 2024-01-11 at 10 55 11

Because it is hard coded to a fixed format:

https://github.com/woocommerce/woocommerce/blob/c487c84778b02946ce811f89368f369d282f1d3e/plugins/woocommerce-blocks/assets/js/blocks/checkout/address-card/index.tsx#L32

### Steps to reproduce

1. Go to the checkout page and select Hungary as the billing country
2. Notice that the first name and last name fields will switch order automatically
3. Refresh the page to see the address card and the name will be in the wrong order

### WordPress Environment

_Not relevant_

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by opr**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

- Adds new utils for parsing address formats, specifically for use in the `AddressCard` component.
  - `getFormattedCountry` - returns the country name, based on its code from a list of countries
  - `getFormattedState`  - returns the state name, based on its code from a list of states
  - `extractName` - finds the "name" portion of an address format and returns it
  - `formatAddress` - takes an address object and a format string and returns the address as an array of strings.

Updates the `AddressCard` component to use the new utils and use [the format defined in `WC_Countries`](https://github.com/woocommerce/woocommerce//blob/a66075ee115c3d699c37240d7a597fa64ed175c5/plugins/woocommerce/includes/class-wc-countries.php)

Closes #42106
Closes #43500

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. As a logged in shopper, add an item to your cart and go to the Checkout block.
2. Fill in your address using an address in Hungary (a zip code that works is `1234`). Take note of the first and last name you use.
3. Place the order then re-add an item and go back to the Checkout block.
4. In the address card you should see your last name then first name. (e.g. if you enter Jon Doe in the checkout, it would display now as Doe Jon)
5. Ensure the format in the address card follows the rest of the Hungarian format as defined in `WC_Countries`. i.e. `{company}, {city}, {address_1}, {address_2}, {postcode}, {country}`
6. Repeat the test using an address in Japan. (A zip code that works is `123-1111`).
7. Return to the Checkout block after placing your order and verify the format is matched again `{postcode}, {state} {city},  {address_1}, {address_2}, {company}, {country}`

#### Developer only test

- Add a snippet to your site:
```php
add_filter( 'woocommerce_localisation_address_formats', function( $formats ) {
	$formats['JP'] = "{company}\n{name}\n{address_1}\n{address_2}\n{city}\n{state}\n{postcode}\n{country}";
    return $formats;
} );
```
Repeat the above testing steps for Japan and verify the format matches what you've entered here, except the name (it will be extracted and shown above).
You can also add custom text between the tags and see if it renders correctly.
Try breaking this by messing with the formatting etc.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/assets/js/blocks/checkout/test/block.js [hunk 1]
```diff
@@ -149,15 +149,39 @@ describe( 'Testing cart', () => {
...
+
 		await waitFor( () => expect( fetchMock ).toHaveBeenCalled() );
 
 		expect(
 			screen.getByRole( 'button', { name: 'Edit address' } )
 		).toBeInTheDocument();
 
 		expect(
-			screen.getByText( 'Ontario', {
+			screen.getByText( 'Toronto ON M4W 1A6', {
```
Comment (opr) [inline plugins/woocommerce-blocks/assets/js/blocks/checkout/test/block.js:160]: This is updated according to the Canadian address formatting

**Review by nielslange — COMMENTED** [review]
Thanks for working on this, @opr.

I noticed that in the developer only test, the return statement is missing.
```php
add_filter( 'woocommerce_localisation_address_formats', function( $formats ) {
	$formats['JP'] = "{company}\n{name}\n{address_1}\n{address_2}\n{city}\n{state}\n{postcode}\n{country}";
} );
```
should be
```php
add_filter( 'woocommerce_localisation_address_formats', function( $formats ) {
	$formats['JP'] = "{company}\n{name}\n{address_1}\n{address_2}\n{city}\n{state}\n{postcode}\n{country}";
	return $formats;
} );
```
Otherwise, em empty address card would be visible:
Screenshot 2024-03-26 at 15 06 49

---

Nitpic: I noticed that the field order of the address card and the billing/shipping address is different. E.g. these are the formats for `HU`:

<table>
<tr>
<td valign="top" width="50%">Address card format:
<br><br>

```
{last_name} {first_name}
{company}
{city}
{address_1}
{address_2}
{postcode}
{country}
```

Screenshot 2024-03-26 at 15 28 50

Screenshot 2024-03-26 at 15 31 16

</td>
<td valign="top" width="50%">Billing/shipping address format:
<br><br>

```
{last_name} {first_name}
{company}
{country}
{postcode} {city}
{address_1}
{address_2}
{state} {phone}
```

Screenshot 2024-03-26 at 15 29 03

</td>
</tr>
</table>

I wonder if we shouldn't ensure that we're using the same format consistently. What are your thoughts on this?

File: plugins/woocommerce-blocks/assets/js/blocks/checkout/test/block.js [hunk 2]
```diff
@@ -149,15 +149,72 @@ describe( 'Testing cart', () => {
...
+		expect(
+			screen.getByText( 'Merseyside', {
+				selector: '.wc-block-components-address-card span',
+			} )
+		).toBeInTheDocument();
+		expect(
+			screen.getByText( 'L1 0BP', {
 				selector: '.wc-block-components-address-card span',
 			} )
 		).toBeInTheDocument();
```
Comment (nielslange) [inline plugins/woocommerce-blocks/assets/js/blocks/checkout/test/block.js:219]: These tests look like we're only checking if the string occurs, but we do not check the expected format. Shouldn't we do something like:

```js
expect(
	screen.getByText( 'Address 1 GB, Liverpool, Merseyside, L1 0BP, United Kingdom (UK)', {
		selector: '.wc-block-components-address-card span',
	} )
).toBeInTheDocument();
```

Or is this not possible due to the way we're rendering these fields? Looking at the source code, I see that we're adding the commas via the `::after` pseudo-element. 

Screenshot 2024-03-26 at 15 44 16

File: plugins/woocommerce-blocks/assets/js/blocks/checkout/utils.ts
Comment (nielslange) [inline plugins/woocommerce-blocks/assets/js/blocks/checkout/utils.ts:1]: What are your thoughts on documenting `extractName` and `formatAddress`? Maybe on https://developer.woo.com/docs/category/code-snippets/. 🤔

File: plugins/woocommerce-blocks/assets/js/blocks/checkout/address-card/index.tsx [hunk 3]
```diff
@@ -28,33 +26,28 @@ const AddressCard = ( {
 	target: string;
 	fieldConfig: FormFieldsConfig;
 } ): JSX.Element | null => {
-	const formattedCountry = isString( ALLOWED_COUNTRIES[ address.country ] )
-		? decodeEntities( ALLOWED_COUNTRIES[ address.country ] )
-		: '';
+	const addressFormats = getSetting< Record< string, string > >(
+		'addressFormats',
+		{}
+	);
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/blocks/checkout/address-card/index.tsx]: This should be part of each country instead of its top level setting.

**Comment by opr** [comment]
@nielslange thanks for your review.

In your comment you mentioned the format and address layout for HU are different.

> I wonder if we shouldn't ensure that we're using the same format consistently. What are your thoughts on this?

In WC Core there are two settings that control this - the format for the address ([controlled here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-countries.php#L563)) vs the [country locale, which contains field priority](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-countries.php#L851), so trying to use one to inform the other is incorrect behaviour.

File: plugins/woocommerce-blocks/assets/js/blocks/checkout/utils.ts
Comment (opr) [reply plugins/woocommerce-blocks/assets/js/blocks/checkout/utils.ts:1]: As these are internal functions only I don't think there is a lot of value documenting these. Third party devs can't use them, or modify their output.

File: plugins/woocommerce-blocks/assets/js/blocks/checkout/address-card/index.tsx (hunk 3 above)
Comment (opr) [reply plugins/woocommerce-blocks/assets/js/blocks/checkout/address-card/index.tsx]: Done, moved `format` into each `countryData` if it has a format, and added `defaultAddressFormat` as a top level setting.

**Comment by nielslange** [comment]
> In WC Core there are two settings that control this - the format for the address ([controlled here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-countries.php#L563)) vs the [country locale, which contains field priority](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-countries.php#L851), so trying to use one to inform the other is incorrect behaviour.

@opr Can you help me understand why applying the same format would be incorrect behaviour? For example, in the address card, we show the city, the two address fields and then the postcode, while in the shipping and billing forms, we show the postcode, the city, and then the two address fields. I'd think that postcode and city should belong together. That was my primary point regarding applying the same format.

**Comment by opr** [comment]
> > In WC Core there are two settings that control this - the format for the address ([controlled here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-countries.php?rgh-link-date=2024-04-02T05%3A13%3A03Z#L563)) vs the [country locale, which contains field priority](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-countries.php?rgh-link-date=2024-04-02T05%3A13%3A03Z#L851), so trying to use one to inform the other is incorrect behaviour.
> 
> @opr Can you help me understand why applying the same format would be incorrect behaviour? For example, in the address card, we show the city, the two address fields and then the postcode, while in the shipping and billing forms, we show the postcode, the city, and then the two address fields. I'd think that postcode and city should belong together. That was my primary point regarding applying the same format.

We spoke about this on a call, and the format is used for _display_ whereas the locale (for form) is used for collection. It may be worth opening an issue to discuss changing the locale to match the format more closely for certain countries.

**Comment by nielslange** [comment]
> We spoke about this on a call, and the format is used for _display_ whereas the locale (for form) is used for collection. It may be worth opening an issue to discuss changing the locale to match the format more closely for certain countries.

I just created [Sync format of address card and the billing/shipping address](https://github.com/woocommerce/woocommerce/discussions/46325) as a discussion item. That way, we can get feedback from the community on whether such a change is desired or not.

**Review by nielslange — APPROVED** [review]
Thanks for working on this, @opr. I've successfully tested the PR.

**Pull request merged**